### DISCUSSION TITLE
run UMEP.ShadowGenerator from python

### DISCUSSION THREAD

**Discussion by s-martiz** [UMEP-dev/UMEP#417] [Python]
Dear UMEP-community,

I'm trying to run UMEP from python to avoid using QGIS and facilitate debugging.
My question is quite simple > Is there a simple example of the shadow_generator.py that I can run from python to better understand how it works ?
Let’s say something like :
ShadowGenerator(geojson_object, position, date, hour) > .tiff object representing the shadow of the zone.

Simon

**Comment by biglimp** [comment]
Hi Simon,
Here is some code to debug the shadow algorithm. This code requires gdal but actually, any 2D numpy array would do as the **dsm** variable. All the functions are available in the UMEP tool.

```
import numpy as np
from osgeo import gdal
import matplotlib.pyplot as plt
from Utilities.SEBESOLWEIGCommonFiles import shadowingfunctions as sh
import time
# import wallalgorithms as wa # if you want shadow on walls. Otherwise, not needed. Just create matrices like np.zeros([rows, cols]). Same # goes for cdsm and tdsm (vegetation) 

filepath_dsm = './data/DSM_KRbig.asc'
filepath_cdsm = './data/CDSM_KRbig.asc'
usevegdem = 0 
trans = 0.05 # transmissivity of shortwave radiation through vegetation
trunkratio = 0.23 # hieght of trunk zone based on cdsm
altitude = 54.8  # sun altitude in degrees
azimuth = 187.7 # sun azimuth in degrees

gdal_dsm = gdal.Open(filepath_dsm)
dsm = gdal_dsm.ReadAsArray().astype(float)
sizex = dsm.shape[0]  # rows
sizey = dsm.shape[1]  # cols
rows = dsm.shape[0]
cols = dsm.shape[1]
geotransform = gdal_dsm.GetGeoTransform()
scale = 1 / geotransform[1]

gdal_cdsm = gdal.Open(filepath_cdsm)
vegdsm = gdal_cdsm.ReadAsArray().astype(float)
gdal_bh = gdal.Open('./data/heightnew_KRbig.tif')
walls = gdal_bh.ReadAsArray().astype(float)
gdal_ba = gdal.Open('./data/aspectnew_KRbig.tif')
wallaspect = gdal_ba.ReadAsArray().astype(float)

vegdsm2 = vegdsm * trunkratio

# walls = wa.findwalls(dsm, 3)



if usevegdem == 1:
    # Vegetation transmittivity of shortwave radiation
    psi = trans
    # amaxvalue
    vegmax = vegdsm.max()
    amaxvalue = dsm.max() - dsm.min()
    amaxvalue = np.maximum(amaxvalue, vegmax)

    # Elevation vegdsms if buildingDEM includes ground heights
    vegdem = vegdsm + dsm
    vegdem[vegdem == dsm] = 0
    vegdem2 = vegdsm2 + dsm
    vegdem2[vegdem2 == dsm] = 0

    # % Bush separation
    bush = np.logical_not((vegdem2 * vegdem)) * vegdem
else:
    psi = 1.
    bush = np.zeros([rows, cols])
    amaxvalue = 0

#f = sh.shadowingfunctionglobalradiation(dsm, azimuth, altitude, scale) # without wall shadows
t0 = time.time()
sh1, wallsh, wallsun, facesh, facesun = sh.shadowingfunction_wallheight_13(dsm, azimuth, altitude, scale, walls, wallaspect*np.pi/180.)
print('Time: ' + str(time.time() - t0))

#  Vegetation included
# vegsh, sh, _, wallsh, wallsun, wallshve, _, facesun = sh.shadowingfunction_wallheight_23(dsm, vegdem, vegdem2, azimuth,
#                                                 altitude, scale, amaxvalue, bush, walls, wallaspect * np.pi / 180.)

plt.imshow(f)
plt.show()
```
Hopefully you get the code to work. /Fredrik

**Reply by s-martiz** [reply]
Thank you for your answer, 
I think theis debug code is not completely updated as I get this error 

from Utilities.SEBESOLWEIGCommonFiles import shadowingfunctions as sh
ImportError: cannot import name 'shadowingfunctions'


Indeed, there is no shadowingfunctions in the SEBESOLWEIGCommonFiles  folder (see below) 
![image](https://user-images.githubusercontent.com/50947346/169473145-2b940916-71a0-4333-b0b2-95758ba4cfe4.png)

Regards
Simon

**Reply by biglimp** [reply]
ok, my misstake. Then just point towards one of the two shadowing functions directly instead

**Reply by s-martiz** [reply]
Thank you very much @biglimp ,
Last question, where can I find a correct test dataset ?
I downloaded the Goteborg_SWEREF99_1200.zip file, but I can't find DSM_KRbig.asc in it 
![image](https://user-images.githubusercontent.com/50947346/169475001-4521447b-4c4e-4350-93ec-1eb149c8f661.png)

**Reply by biglimp** [reply]
Just change the code to read the .tif-files instead (DSM_KRbig.tif)

**Reply by bbrangeo** [reply]
filepath_dsm = './data/DSM_KRbig.tif'

**Reply by bbrangeo** [reply]
from Utilities.SEBESOLWEIGCommonFiles import shadowingfunction_wallheight_13


...

sh1, wallsh, wallsun, facesh, facesun = shadowingfunction_wallheight_13(dsm, azimuth, altitude, scale, walls, wallaspect*np.pi/180.)

**Reply by s-martiz** [reply]
Thanks @bbrangeo and @biglimp 
Still, I can't find this files in the Goteborg example
> heightnew_KRbig.tif
> aspectnew_KRbig.tif
![image](https://user-images.githubusercontent.com/50947346/169521907-1086d3ea-9662-4b8c-bcce-204359fc86a4.png)

**Reply by biglimp** [reply]
They are not included. Create them in Wall and height aspect tool in UMEP.

**Reply by s-martiz** [reply]
I finally manage to make it works
thank you @biglimp 
![image](https://user-images.githubusercontent.com/50947346/169539223-d4f65450-3d9d-4192-a74d-c2a26c47b45f.png)
